### PR TITLE
Handle week=5 MWD cases in posix TZ strings

### PR DESCRIPTION
**Pull request by Manishearth**
Our calculations were incorrect for week=5 cases (which are the "last $weekday of the month", not the fifth.

I see a todo around WithLeap/NoLeap, however there are only two timezones that use it (Africa/Casablanca, Africa/El_Aiun), and they both do a transition from NoLeap(0) to WithLeap(365) which seems to mean that they have a very short period of DST between two years. That seems confusing. Their posix tz string does not kick in until 2087.

Also took the chance to fix a bug in transitions calculations I discovered.

### CODE REVIEW

**Comment by Manishearth** [comment]
Selected data for `Africa/Casablanca` (`Africa/El_Aiun` is similar):

```
...
transitions[194]          3699828000           2     2087-03-30T02:00:00+00:00[Africa/Casablanca]
transitions[195]          3703456800           1     2087-05-11T03:00:00+01:00[Africa/Casablanca]

...

--------- Footer ----------
std_info                  XXX                  -7200 02:00:00
dst_info                  
  variant_info            +01                  -3600 01:00:00
  start_date              WithLeap(0): Jan 1   0     00:00:00
  end_date                NoLeap(365): Jan 1   82800 23:00:00

```

Note that our code assumes that timezones either both use WithLeap or both use NoLeap, so the code won't be running anyway.

The actual sources online, and tzdb, don't have any such autoDST. This is perhaps bogus data, inserted here for some reason?

File: src/tzdb.rs [hunk 1]
```diff
@@ -401,7 +401,15 @@ impl Tzif {
             }
         };
 
-        let year = utils::epoch_time_to_iso_year(epoch_seconds.0 * 1000);
+        // Calculate year, but clamp it to the last transition
+        // We do not want to try and apply the posix string to earlier years!
+        let mut epoch_seconds_for_year_calculation = epoch_seconds;
```
Comment (nekevss) [inline src/tzdb.rs:409]: thought: maybe add a "re: Antarctica/Troll" or something similar to link to a specific zone

**Comment by nekevss** [comment]
> The actual sources online, and tzdb, don't have any such autoDST. This is perhaps bogus data, inserted here for some reason?

`Africa/Casablanca` is Morocco, which has Ramadan DST. That's probably where the extended calculations come from ... but I'd have to double check.

File: src/builtins/compiled/zoneddatetime.rs [hunk 2]
```diff
@@ -678,6 +692,11 @@ mod tests {
         (LONDON_TRANSITION_1968_02_18_MINUS_ONE, true),
         (SAMOA_IDL_CHANGE, true),
         (SAMOA_IDL_CHANGE_MINUS_ONE, true),
+        (LONDON_POSIX_TRANSITION_2019_03_31_MINUS_ONE, true),
+        (LONDON_POSIX_TRANSITION_2019_03_31, true),
+        (LONDON_POSIX_TRANSITION_2017_03_26_MINUS_ONE, true),
+        (LONDON_POSIX_TRANSITION_2017_03_26_MINUS_ONE, true),
+        (TROLL_FIRST_TRANSITION, true),
```
Comment (nekevss) [inline src/builtins/compiled/zoneddatetime.rs:699]: thought: I wonder if everything here is fine with `Europe/Riga` too. I think there's some test cases here that cover it. But that was another zone with a transition edge case that I ran into.

File: src/utils.rs [hunk 3]
```diff
@@ -48,43 +48,34 @@ pub(crate) fn epoch_ms_to_ms_in_day(t: i64) -> u32 {
     (t.rem_euclid(i64::from(MS_PER_DAY))) as u32
 }
 
-/// Mathematically determine the days in a year.
-pub(crate) fn mathematical_days_in_year(y: i32) -> i32 {
+pub(crate) fn is_leap(y: i32) -> bool {
```
Comment (nekevss) [inline src/utils.rs:52]: note: the slightly odd util names comes from the Temporal specification, mainly [Section 13.3](https://tc39.es/proposal-temporal/#sec-date-equations). I'm not against cleaning the names, but it may be worth commenting the date equation these functions are associated with.

File: src/builtins/compiled/zoneddatetime.rs (hunk 2 above)
Comment (Manishearth) [reply src/builtins/compiled/zoneddatetime.rs:699]: I've filed https://github.com/boa-dev/temporal/issues/500 for this
